mean.hhs

```diff
@@ -0,0 +1,104 @@
+/**
+ * @author Jianan Lin (林家南)
+ * @param input - the list, matrix, tensor or structure that you want to get the mean(average) of
+ * @returns - a number representing the mean of all the elements of a list, matrix, tensor or other structure
+ * 
+ */
+
+
+*import math: is_number
+*import math: flatten
+*import math: ndim
+*import math: deep_copy
+
+
+function mean(input) {
+
+    // if there are no arguments, there is an error
+    if (arguments.length === 0) {
+        throw new Error('No argument given');
+    }
+
+    // if the first argument is a number, then we should get the mean of a list of numbers
+    if (is_number(arguments[0])) {
+
+        // determine whether there is invalid element
+        for (let i = 0; i < arguments.length; i++) {
+            if (!is_number(arguments[i])) {
+                throw new Error('First argument suggests a list, but there is a non number element in arguments.');
+            }
+        }
+
+        // sum_array is the sum, n is the number, result is sum / n
+        let sum_array = 0, n = 0;
+        for (let i = 0; i < arguments.length; i++) {
+            sum_array += arguments[i];
+            n++;
+        }
+
+        let result = sum_array / n;
+        return result;
+    }
+
+    //now we're dealing with non-numbers and matrix-like structures:
+    //declare raw_in, set it to a input clone if its a Mat otherwise input itself
+    let in_type = (input instanceof Mat);
+    let raw_in = (in_type) ? input.clone() : deep_copy(input);
+
+    // if the input is a matrix-like structure, then should we add another argument to decide : 1. get the mean of the whole matrix; 2. get the mean of each row / column of the matrix by axis?
+    if (arguments.length === 1 && (Array.isArray(arguments[0]) || arguments[0] instanceof Mat)) {
```

Comment:
`if (arguments.length === 1 && (Array.isArray(arguments[0]) **&&ndim(arguments[0] < 3 ** || arguments[0] instanceof Mat))` because it needs to be a 2dimensional at most array, so the next block statement can handle higher dimensional arrays, otherwise this one gets satisfied and doesn't do what it needs to for 3d arrays.

Reply:
I have some ideas. First, I move "arguments.length === 1" in front of these cases (just behind the case "isnumber(arguments[0])". Since at this time we only need to consider one argument, we can throw error if it is not obeyed. Second, I let raw_in = (in_type) ? input.clone().val : deep_copy(input). Of course in_type is true if input is tensor. In this way, we do not need to consider what if raw_in is not a JS array.